### PR TITLE
Allow panning while moving a node

### LINKED ISSUE #27300: Enable panning with middle mouse button while dragging a node.
**Godot version:**
3.1 Stable

In larger 2D scenes, having to drag a node large distances gets quite annoying as you must:

1. Drag the node to the edge of the screen
2. Pan to make room for more dragging
3. Repeat until you have dragged the node to the desired location

And I don't see much reason as to why this shouldn't be a thing.

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes #27300

Also supports simple spacebar panning, which makes it work especially well.

### CODE REVIEW

**Comment by groud** [comment]
Hmm, I think we should likely decouple the panning from the dragging in general, having a DRAG_PAN_WITH_MOVE drag type kinda looks like a dirty workaround.

I would likely add a boolean to check whether we are dragging or not, since all other modes should still allow for panning while dragging.

**Comment by KoBeWi** [comment]
I made an update. Now every tool supports panning, including anchors and selection.

Although it's worth to investigate if changing `return true` to return `!panning` has any side effects. I changed it, so that other action's event accepting doesn't interfere with panning.

Now that I think of it, it would be simpler to just move panning action to the top. Doing that however disables the ability to move something continuously with cursor's wrap.

**Comment by groud** [comment]
It looks kinda better now, but I don't think all those `return !panning` are good, we should keep the `return true ` as it means that the event was handled by the function.

You should move the dedicated `_gui_input_zoom_or_pan()` handler in `_gui_input_viewport()`, so that it gets executed whether or not the event was accepted elsewhere. 
Also, you likely need to keep the DRAG_PAN value, since it is still useful for the pan tool I think.

**Comment by groud** [comment]
Awesome, it looks a lot better right now :)
Could you [squash the three commits together](https://godot.readthedocs.io/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) ?

**Comment by KoBeWi** [comment]
> You should move the dedicated _gui_input_zoom_or_pan() handler in _gui_input_viewport()

I totally didn't thing of that D: `_gui_input_zoom_or_pan()` is now out of the if/else chain.

> Also, you likely need to keep the DRAG_PAN value, since it is still useful for the pan tool I think.

Nah, this is the only line that uses it (aside from the one that changes cursor):

`(b->get_button_index() == BUTTON_LEFT && tool == TOOL_PAN) ||`

It's already handled by `panning = true`, so DRAG_PAN is unnecessary now.

**Comment by groud** [comment]
> It's already handled by panning = true, so DRAG_PAN is unnecessary now.

Yeah I guess it makes sense :)

**Comment by KoBeWi** [comment]
I also added initialization for `panning` in the constructor (since all variables have this) and squashed the commits.

**Review by groud — APPROVED** [review]

**Comment by avencherus** [comment]
Couldn't get it to pan while dragging and then holding space.  The middle mouse button one appears to engage it (though difficult to click and hold both), but it does have some wrapping side effect when you have the drag near the edges.

Is it supposed to do this?

ezgif com-video-to-gif (2)

**Comment by KoBeWi** [comment]
> Couldn't get it to pan while dragging and then holding space.

That's because you need to enable "simple space panning" in Editor Settings. Although I guess this could get "enabled" by default when dragging 🤔

> wrapping side effect

Normal panning also supports wrapping, so now with dragging it's even more useful.

**Comment by KoBeWi** [comment]
I just added a commit that allows spacebar panning with simple spacebar panning is disabled BUT while dragging something (figured it makes sense, see 2 above comments).

Also .-.
While testing the standard spacebar panning (Space + Click), I noticed that it might unexpectedly move things when holding space bar (I mean, hold space and try to move, but it can also activate dragging). Is this important thing to fix? >.> It's independent from the commit I just made.

**Comment by avencherus** [comment]
Working nicely now, makes things much easier.  👍 

That jittering of the dragged item I noticed as well, would it be something like a frame off timing with the rendering?

It looks like the icon is getting a little bit of the motion of the panning, then snapping back to place, so maybe not any actual motion being applied to it, but rather its still at its previous position for a frame.

A lot of yields and such seem to be in the UI code, and I tend to bump into a lot of intuitive issues where properties and such are not immediately available after some action.

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 1]
```diff
@@ -2259,8 +2259,6 @@ void CanvasItemEditor::_gui_input_viewport(const Ref<InputEvent> &p_event) {
 		case DRAG_MOVE:
 			c = CURSOR_MOVE;
 			break;
-		case DRAG_PAN:
-			c = CURSOR_DRAG;
```
Comment (akien-mga) [inline editor/plugins/canvas_item_editor_plugin.cpp]: Needs rebase as I added a `break:` here, so there's a merge conflict.

Comment (akien-mga) [reply editor/plugins/canvas_item_editor_plugin.cpp]: Please also squash the commits together.

Comment (KoBeWi) [reply editor/plugins/canvas_item_editor_plugin.cpp]: This merge messed up everything again. I just give up .-.
I'll make a new PR soon.

**Pull request closed without merging**